scroll-verifier/contracts/ScrollVerifier.sol

```
@@ -0,0 +1,82 @@
+import {RLPReader} from "@eth-optimism/contracts-bedrock/src/libraries/rlp/RLPReader.sol";
+
+
+interface IScrollChain {
+    /// @param batchIndex The index of the batch.
+    /// @return The state root of a committed batch.
+    function finalizedStateRoots(uint256 batchIndex) external view returns (bytes32);
+}
+
+interface IScrollChainCommitmentVerifier {
+    function verifyZkTrieProof(
+        address account,
+        bytes32 storageKey,
+        bytes calldata proof
+    ) external view returns (bytes32 stateRoot, bytes32 storageValue);
+
+    function rollup() external view returns (address);
+
+    function verifyStateCommitment(
+        uint256 batchIndex,
+        address account,
+        bytes32 storageKey,
+        bytes calldata proof
+    ) external view returns (bytes32 storageValue);
+}
+
+struct ScrollWitnessData {
+    uint256 batchIndex;
+}
+
+struct StateProof {
+    bytes[] storageProofs;          // An array of proofs of individual storage elements 
+}
+
+contract ScrollVerifier is IEVMVerifier {
+    error InvalidSlotSize(uint256 size);
+    IScrollChainCommitmentVerifier public immutable verifier;
+    string[] _gatewayURLs;
+
+    constructor(string[] memory _urls, IScrollChainCommitmentVerifier _verifierAddress) {
+        verifier = _verifierAddress;
+        _gatewayURLs = _urls;
+    }
+
+    /*
+     * Retrieves an array of gateway URLs used by the contract.
+     * @returns {string[]} An array containing the gateway URLs.
+     *     */
+    function gatewayURLs() external view returns (string[] memory) {
+        return _gatewayURLs;
+    }
+
+
+    function getTrieProof(address target, uint256 slot, bytes memory compressedProof, bytes32 root) internal view returns(bytes memory){
+        (bytes32 stateRoot, bytes32 storageValue) = verifier.verifyZkTrieProof(target, bytes32(slot), compressedProof);
+        require(stateRoot == root, "stateRoot not matched");
```

This should throw an error object.